### ISSUE TITLE
Indentation errors within ternary expressions

### ISSUE THREAD

**Issue by mjethani** [eslint/eslint#8977]
**Tell us about your environment**

* **ESLint Version:** v4.2.0
* **Node Version:** v6.10.2
* **npm Version:** 3.10.10

**What parser (default, Babel-ESLint, etc.) are you using?**

Default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```json
{
  "env": {
    "browser": true,
    "es6": true
  },
  "extends": "eslint:recommended",
  "parserOptions": {
    "sourceType": "module"
  },
  "rules": {
    "indent": [
      "error",
      2
    ],
    "linebreak-style": [
      "error",
      "unix"
    ],
    "quotes": [
      "error",
      "double"
    ],
    "semi": [
      "error",
      "always"
    ],
    "no-unused-vars": [
      "off"
    ],
    "no-constant-condition": [
      "off"
    ]
  }
}
```

</details>

**What did you do? Please include the actual source code causing the issue.**

```js

export function foo()
{
  const bar = true ?
                true :
                false;
  return true &&
         (true ?
            true :
            false);
}
```

**What did you expect to happen?**

No errors.

**What actually happened? Please include the actual, raw output from ESLint.**

```
/Users/mjethani/tmp/eslint/foo.js
  4:1  error  Expected indentation of 4 spaces but found 16   indent
  5:1  error  Expected indentation of 4 spaces but found 16   indent
  8:1  error  Expected indentation of 11 spaces but found 12  indent
  9:1  error  Expected indentation of 11 spaces but found 12  indent

✖ 4 problems (4 errors, 0 warnings)
  4 errors, 0 warnings potentially fixable with the `--fix` option.
```

**Additional notes**

This appears to be a regression in ESLint 4.0 as it was working perfectly fine in ESLint 3.19.

**Comment by not-an-aardvark** [comment]
Hi, thanks for reporting.

This is working as intended. In ESLint 3.x, the `indent` rule completely ignored the indentation of ternary expressions. Starting in ESLint 4.x, it now verifies that they are indented correctly. In your case, the rule expects an indentation of 4 spaces on lines 4 and 5 because this is at an offset of 2 from the start of the variable declaration, and the rule expects an indentation of 11 spaces on lines 8 and 9 because it's offset by 2 spaces from the `(true` on line 7.

If you would like to replicate the 3.x behavior, you can enable the `indent-legacy` rule instead of the `indent` rule, which is a copy of the `indent` rule from 3.x.

**Comment by mjethani** [comment]
Fair enough, but I'd say it's a missing feature. There should be an option to specify the indentation level within ternary expressions. It would help us enforce our [coding style](https://adblockplus.org/coding-style).

Thanks for the pointer to `indent-legacy`.

**Comment by not-an-aardvark** [comment]
Could you elaborate on how you would expect the option to behave?

**Comment by mjethani** [comment]
It could be called `TernaryExpression` and it could take either a numeric value or the value `"off"`.

With `TernaryExpression` set to `1` (default), this would be correct:

```js
const foo = veryLongName ?
              anotherVeryLongName :
              yetAnotherVeryLongName;
```
This too would be correct:

```js
function foo()
{
  return somethingThatMatters &&
         (veryLongName ?
            anotherVeryLongName :
            yetAnotherVeryLongName);
}
```

**Comment by platinumazure** [comment]
@mjethani I think the `indent` rule usually defines a `1` result as being offset from the first token of the line:

```js
// Using ConditionalExpression as the ESTree node name, but probably doesn't have to be set in stone
/* eslint indent: ["error", 2, { "ConditionalExpression": 1 }] */
const foo = veryLongName ?
  anotherVeryLongName :
  yetAnotherVeryLongName;
```

**Comment by not-an-aardvark** [comment]
The reason for the current behavior is because aligning from `veryLongName` is inconsistent with how indentation works for other code. For example, people typically don't offset array elements from the opening bracket:

```js
foo = [
        1,
        2,
        3
      ];
```

Instead, they offset the elements from the start of the line:

```js
foo = [
  1,
  2,
  3
];
```

**Comment by mjethani** [comment]
I see.

In my opinion, for `indent: ["error", 2, {CallExpression: {arguments: 2}]`, most people would find this to be correct:

```js
function foo()
{
  bar(baz(veryLongName, anotherVeryLongName,
          yetAnotherVeryLongName),
      qux(veryLongName, anotherVeryLongName,
          yetAnotherVeryLongName));
}
```

Of course this gives an error.

This works instead:

```js
function foo()
{
  bar(baz(veryLongName, anotherVeryLongName,
      yetAnotherVeryLongName),
  qux(veryLongName, anotherVeryLongName,
      yetAnotherVeryLongName));
}
```

Which is wrong, because `qux(…)` is actually an argument to `bar`.

**Comment by platinumazure** [comment]
@mjethani In your above CallExpression examples, that is achieved using `CallExpression: "first"`.

**Comment by mjethani** [comment]
@platinumazure That is not what I'm trying to achieve. Let's say it's `CallExpression: {arguments: 3}`, then I would expect this to be correct:

```js
function foo()
{
  bar(baz(veryLongName, anotherVeryLongName,
            yetAnotherVeryLongName),
        qux(veryLongName, anotherVeryLongName,
              yetAnotherVeryLongName));
}
```

Instead, `qux` must still end up aligned with `bar`, which is wrong.

I think ESLint should count indentation not from the beginning of the _line_ but from the beginning of the _expression_.

**Comment by not-an-aardvark** [comment]
Would you also want your array to be indented like this? This would be the result of indenting from the beginning of the expression.

```js
foo = [
        1,
        2,
        3
      ];
```

(I'm not asking that to be argumentative -- I'm just trying to determine the natural results of indenting from the beginning of an expression. I've found that indenting from the start of the line is generally more aligned with how people expect indentation to work due to many cases like that array, but I'm interested to see if you think differently.)

**Comment by not-an-aardvark** [comment]
>```js
>// [correct code]
>function foo()
>{
>  bar(baz(veryLongName, anotherVeryLongName,
>      yetAnotherVeryLongName),
>  qux(veryLongName, anotherVeryLongName,
>      yetAnotherVeryLongName));
>}
>```

I agree that this doesn't seem ideal. The reason the rule behaves this way is because people tend to do things like this:

```js
foo = [{
  a: 1
},
{
  b: 2
},
{
  c: 3
}]
```

So the rule aligns all of the elements of the array with the first one, which accounts for this case. It might be possible to tweak the rule a bit to handle your case better without breaking the other common case, but in general there are some places where we can't realistically support everyone's preferences.

**Comment by mjethani** [comment]
I can see why it's done that way.

Let's examine a few cases with `CallExpression: {arguments: "first"}, ArrayExpression: 1` in ESLint 4:

This is correct:

```js
function foo()
{
  bar(baz(veryLongName, anotherVeryLongName,
          yetAnotherVeryLongName, [
            1
          ]));
}
```

But if we lose the third argument and move the opening bracket to the previous line, then this is correct:

```js
function foo()
{
  bar(baz(veryLongName, anotherVeryLongName, [
    1
  ]));
}
```

There is some inconsistency here.

If we throw in a fourth argument, then this is correct:

```js
function foo()
{
  bar(baz(veryLongName, anotherVeryLongName, [
    1
  ],
          yetAnotherVeryLongName));
}
```

(I have a feeling that this one is a bug.)

Anyway, it looks to me like for some expressions it should be from the beginning of the line (e.g. arrays), while for others it should be from the beginning of the expression (e.g. function calls). In the case of a ternary expression, if it breaks into multiple lines, I don't see why it should be indented from the beginning of the line and not from the beginning of the expression itself.

Now I understand that this might actually be super difficult to implement given that a ternary expression itself can consist of multiple nested expressions, but it's something to think about.

**Comment by wojdyr** [comment]
It seems that ternary expressions have particularly many indentation styles. Maybe it'd be worth to add an option to ignore the indentation of ternary expressions? (as it was in ESLint 3.x)

**Comment by not-an-aardvark** [comment]
That will be addressed by https://github.com/eslint/eslint/issues/8594.

**Comment by Kovensky** [comment]
I ran into a similar problem that I am not sure is related to this or not.

```js
const obj = {
  foo () {
    return condition ? // comment
    1 :
      2
  }
}
```

Removing the `// comment` after the `?` makes the rule correctly indent `1` at the same level as `2`.

**Comment by not-an-aardvark** [comment]
@Kovensky I don't think that's related. Can you open a new issue?

**Comment by Kovensky** [comment]
#9729 done

**Comment by not-an-aardvark** [comment]
Closing because this is working as intended.
